text/0000-update-version-handling.md

```diff
@@ -0,0 +1,225 @@
+
+| Original                     | Proposed                     |
+| ---------------------------- | ---------------------------- |
+| `notion install node latest` | `notion install node@latest` |
+| `notion pin node 10`         | `notion pin node@10`         |
+| `notion fetch node 10.5`     | `notion fetch node@10.5`     |
+
+This also works for shorthands (currently `latest`, conceivably also tags like `next` or `lts` in the future):
+
+| Original                 | Proposed                  | Status      |
+| ------------------------ | ------------------------ | ------------ |
+| `notion pin node latest` | `notion pin node@latest` | current      |
+| `notion pin node lts`    | `notion pin node@lts`    | current      |
+| `notion pin node next`   | `notion pin node@next`   | hypothetical |
+
+## Installing, fetching, or pinning multiple tools
+
+Previously, installing multiple tools required invoking Notion multiple times. Now, it may be invoked with multiple tools at once.
+
+- **Original:**
+
+    ```ts
+    notion install create-react-app latest
+    notion install typescript 3.4
+    ```
+
+- **Updated:**
+
+    ```
+    notion install create-react-app@latest typescript@3.4
+    ```
+
+There are no impacts to features other than fetching, installing, and pinning.
+
+## Implementation changes
+
+The command line implementation will need to be updated, with the install, fetch, and pin commands now receiving a list of tool specs.
+
+The command line parsing will need to be updated to accept variadic arguments of tool-with-optional-versions, using Clap's [`Arg::multiple`](https://docs.rs/clap/2.32.0/clap/struct.Arg.html#method.multiple) argument, with a custom parser. E.g., for `install`, we might have something like this
+
+```rust
+#[derive(StructOpt)]
+pub(crate) struct Install {
+    /// The tool to install, e.g. `node` or `npm` or `yarn`
+    #[structopt(multiple = true), parse(try_from_str = "ToolSpec::try_from_str")]
+    tools: Vec<ToolSpec>,
+}
+```
+
+Here we assume that the `ToolSpec` will add a `try_from_str` or `try_from_os_str` implementation, with a useful error type—here hand-waved as `String`, but it could be as complex as we need as long as the error type has an implementation of `Into<String>`:
+
+```rust
+impl ToolSpec {
+  fn try_from_str(string: &str) -> Result<ToolSpec, String> {
+    // split apart the tool name and its version
+  }
+}
+```
+
+This is a relatively simple implementation. The only complication is the existence of Node namespaces: `@namespace/package`. This means parsing the version is not as simple as splitting on `"@"`. However, it should still be straightforward using the combination of a regular expression and the existing semver tooling.
+
+The `ToolSpec` change will make the format `tool[@version]` work; updating the commands to take a `Vec<ToolSpec>` will support passing multiple arguments.
```

Comment:
One important implementation detail (doesn't need to be added to the RFC itself, just so that we don't forget): We'll need to make sure when doing multiple `install` / `pin` operations that we do the `node` one (if present) _first_, since that sets up the environment.

Reply:
This was a great callout, I implemented it [here](https://github.com/notion-cli/notion/pull/383/commits/e045d32ef174cf3cd1a102a8ade01f1dfdecacdb). Being able to just `#[derive(PartialEq, Eq, PartialOrd, Ord)]` on that enum and have it work automatically was ✨ magic ✨.